addon/-private/class/modifier.ts

```diff
@@ -82,14 +76,51 @@ export default class ClassBasedModifier<S = DefaultSignature> {
     setOwner(this, owner);
     this.args = args;
 
-    deprecateForDestroyables('willDestroy', this);
-    deprecateForDestroyables('isDestroying', this);
-    deprecateForDestroyables('isDestroyed', this);
-
     assert(
       'ember-modifier: You cannot implement both `modify` and any of the deprecated legacy lifecycle hooks (`didInstall`, `didReceiveArguments`, and `didUpdateArguments`)',
       !(_implementsModify(this) && _implementsLegacyHooks(this))
     );
+
+    deprecate(
+      `Modifier.${name} is deprecated. Use the corresponding API from '@ember/destroyable' instead.`,
+      (['willDestroy', 'isDestroying', 'isDestroyed'] as const).some(
+        (name) => this[name] !== ClassBasedModifier.prototype[name]
+      ) && !SEEN_CLASSES_FOR_DESTROYABLES.has(this.constructor),
+      {
+        id: 'ember-modifier.use-destroyables',
+        until: '4.0.0',
+        for: 'ember-modifier',
+        since: {
+          available: '3.2.0',
+          enabled: '3.2.0',
+        },
+      }
+    );
+
+    if (DEBUG && !SEEN_CLASSES_FOR_DESTROYABLES.has(this.constructor)) {
+      SEEN_CLASSES_FOR_DESTROYABLES.add(this.constructor);
+    }
+
+    deprecate(
+      `ember-modifier (in ${this.constructor.name} at ${
+        new Error().stack
+      }): The \`didInstall\`, \`didReceiveArguments\`, and \`didUpdateArguments\` hooks are deprecated. Use the new \`modify\` hook instead.`,
```

Comment:
Should we add error stack info to the `willDestroy`/`isDestroying`/`isDestroyed` ones too? Seems generally good, but if you do want to do that I would **not** try to share the stack; I can't think of a way to share that would avoid eagerly computing the `.stack` property (which is pretty expensive) even when you might not trigger a deprecation.

Reply:
Yes, I think so, mean to do so!